### PR TITLE
Opscenter 87: Support dynamic grouped label 

### PR DESCRIPTION
**Pull request by jinfengni**

### CODE REVIEW

**Comment by jinfengni** [comment]
This pr depends on https://github.com/sundeck-io/OpsCenter/pull/227.   The 3rd commit (and follow on) has the code change for this PR.

There are couple of TODOs:
- add test for the modified / new stored procedures.
- add cypress test for the UI change

Tested the basic functionality. Add a dynamic grouped label using "query_type" as the condition expression. 

```
select * from internal.labels where is_dynamic;
+------+------------------+------------+-------------------------+------------+---------+-------------------------+------------+
| NAME | GROUP_NAME       | GROUP_RANK | LABEL_CREATED_AT        | CONDITION  | ENABLED | LABEL_MODIFIED_AT       | IS_DYNAMIC |
|------+------------------+------------+-------------------------+------------+---------+-------------------------+------------|
| NULL | my_dynamic_group |       NULL | 2023-09-04 23:32:34.723 | query_type | NULL    | 2023-09-04 23:32:34.723 | True       |
+------+------------------+------------+-------------------------+------------+---------+-------------------------+------------+

```

Here is the dynamic group "my_dynamic_group" is used to show the queries:

Screen Shot 2023-09-04 at 11 48 37 PM

**Comment by rymurr** [comment]
Looks great @jinfengni !

Will wait for the rebase before a full review. One question, how do we define more complex functions than just eg `query_type`? What if I wanted the qtag source for every query (there is a function in `tools` schema to extract the source)? Or more complicated: qtag source if a normal query, 'INTERNAL' if its a client generated statement

File: bootstrap/004_labels.sql
Comment (rymurr) [inline bootstrap/004_labels.sql:129]: I would add the `if label_condition is not null label_condition else 'Other' AS group_name` here so we always have an 'other' bucket

Comment (jinfengni) [reply bootstrap/004_labels.sql:129]: Makes sense. I'll add 'other' bucket to the group.

**Comment by jinfengni** [comment]
@rymurr , I rebased the PR after #227 was merged.

There is still work need done to handle delete the dynamic group.  The create / update seems to work fine.

Regarding the `TOOLS.TAG_SOURCES()` function used in the condition expression,  seems its returned result is not a string type, and it does not include the value of 'source' key.

For instance, for a query that the broker added the qtags. Here is the `QTAG()` function output:

```
select tools.qtag(query_text), query_type from snowflake.account_usage.query_history where query_text ilike '%"app":"sundeck"%' limit 1;
+----------------------------------------------------+------------+
| TOOLS.QTAG(QUERY_TEXT)                             | QUERY_TYPE |
|----------------------------------------------------+------------|
| [                                                  | SELECT     |
|   {                                                |            |
|     "KEY": "app",                                  |            |
|     "SOURCE": "sundeck",                           |            |
|     "TYPE": "APPLICATION",                         |            |
|     "VALUE": "sundeck"                             |            |
|   },                                               |            |
|   {                                                |            |
|     "KEY": "flow_name",                            |            |
|     "SOURCE": "sundeck",                           |            |
|     "TYPE": "DIMENSION",                           |            |
|     "VALUE": "FLOW_WITH_PRIVATE_RESULTS_TURNED_ON" |            |
|   },                                               |            |
|   {                                                |            |
|     "KEY": "hooks_applied",                        |            |
|     "SOURCE": "sundeck",                           |            |
|     "TYPE": "TRACE",                               |            |
|     "VALUE": []                                    |            |
|   },                                               |            |
|   {                                                |            |
|     "KEY": "query_rejected",                       |            |
|     "SOURCE": "sundeck",                           |            |
|     "TYPE": "DIMENSION",                           |            |
|     "VALUE": false                                 |            |
|   },                                               |            |
|   {                                                |            |
|     "KEY": "query_text_updated",                   |            |
|     "SOURCE": "sundeck",                           |            |
|     "TYPE": "DIMENSION",                           |            |
|     "VALUE": false                                 |            |
|   },                                               |            |
|   {                                                |            |
|     "KEY": "reject_message",                       |            |
|     "SOURCE": "sundeck",                           |            |
|     "TYPE": "TRACE",                               |            |
|     "VALUE": ""                                    |            |
|   },                                               |            |
|   {                                                |            |
|     "KEY": "warehouse_updated",                    |            |
|     "SOURCE": "sundeck",                           |            |
|     "TYPE": "DIMENSION",                           |            |
|     "VALUE": false                                 |            |
|   }                                                |            |
| ]                                                  |            |
+----------------------------------------------------+------------+
```

The `QTAG` function return an array of object.  If we use `QTAG_SOURCES` over the array, it would return an array of index.
```
select tools.qtag_sources(tools.qtag(query_text)), query_type from snowflake.account_usage.query_history where query_text ilike '%"app":"sundeck"%' limit 1;
+--------------------------------------------+------------+
| TOOLS.QTAG_SOURCES(TOOLS.QTAG(QUERY_TEXT)) | QUERY_TYPE |
|--------------------------------------------+------------|
| [                                          | SELECT     |
|   "0",                                     |            |
|   "1",                                     |            |
|   "2",                                     |            |
|   "3",                                     |            |
|   "4",                                     |            |
|   "5",                                     |            |
|   "6"                                      |            |
| ]                                          |            |
+--------------------------------------------+------------+
```

Not sure what's the correct way to use `QTAG_SOURCES` function in the dynamic grouped label condition.

**Comment by rymurr** [comment]
@jinfengni I would simply use `tools.qtag_sources(qtag_filter)[0]`. This will give the first qtag source or `NULL`. Some queries have multiple qtags but I am assuming the `rank` of the grouped labels to be the order of the qtags as they appear.

**Comment by rymurr** [comment]
> There is still work need done to handle delete the dynamic group. The create / update seems to work fine.

what is different about dynamic labels that delete and update are different?

**Comment by jinfengni** [comment]
> what is different about dynamic labels that delete and update are different?

Update for dynamic labels are different: 1) the validation of input is different, 2) the update should uses 'group_name` as the search condition in WHERE clause.  As such, I need pass the flag `is_dynamic` from UI to the stored procedure.

**Comment by jinfengni** [comment]
> Looked at the DDL for `qtag_sources()`. It calls `return Object.keys(QTAG);` . Looks like it will only return the object's field key, not object's field value.

Finally figure out a way to extract the "SOURCE" value.  `tools.qtag(query_text)[0]['SOURCE']` will do the job.

```
select tools.qtag(query_text)[0]['SOURCE'], query_text from snowflake.account_usage.query_history where query_text ilike '%"app":"sundeck"%' limit 1;
+-------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| TOOLS.QTAG(QUERY_TEXT)[0]['SOURCE'] | QUERY_TEXT                                                                                                                                                                                              |
|-------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| "sundeck"                           | copy into orders_jdbc                                                                                                                                                                                   |
|                                     | -- {"app":"sundeck","flow_name":"FLOW_WITH_PRIVATE_RESULTS_TURNED_ON","hooks_applied":[],"query_rejected":false,"query_text_updated":false,"reject_message":"","warehouse_updated":false,"workload":""} |
+-------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

**Comment by jinfengni** [comment]
@rymurr , I rebased and pushed some fix.  The flow of create / update / delete dynamic grouped labels work fine now. I also add couple of new unit tests for the stored procedures.

The Cypress tests need some work.  If you have time, please review the code changes in stored procedures and UI. I'll continue with the Cypress test work.  Thanks!

**Review by rymurr — APPROVED** [review]
Looks great!

**Review by rymurr — APPROVED** [review]

**Comment by jinfengni** [comment]
@rymurr ,  finally got a green run of Cypress test, after getting help from @vqmarkman  on how to debug the Cypress test failure.

If you have time, please take another look. Thanks.

**Comment by rymurr** [comment]
> @rymurr ,  finally got a green run of Cypress test, after getting help from @vqmarkman  on how to debug the Cypress test failure.
> 
> If you have time, please take another look. Thanks. 

I've approved! Thanks @jinfengni

**Pull request merged**
